src/cargo/core/manifest.rs

```diff
@@ -531,6 +531,84 @@ impl Manifest {
         Ok(())
     }
 
+    pub fn check_rust_version(&self) -> CargoResult<()> {
+        let rust_version = match &self.rust_version {
+            Some(v) => v,
+            None => return Ok(()),
+        };
+        let req = match semver::VersionReq::parse(rust_version) {
+            // Exclude semver operators like `^` and pre-release identifiers
+            Ok(req) if rust_version.chars().all(|c| c.is_ascii_digit() || c == '.') => req,
+            _ => bail!("`rust-version` must be a value like \"1.32\""),
+        };
+
+        let check = |major, minor, message| {
+            let unsupported = Version::new(major, minor - 1, 9999);
+            if req.matches(&unsupported) {
+                bail!(
+                    "rust-version `{}` is older than the first version required for {}\n\
+                     This requires a version of at least `{}.{}`.",
+                    rust_version,
+                    message,
+                    major,
+                    minor
+                );
+            }
+            Ok(())
+        };
+
+        if let Some(first_version) = self.edition.first_version() {
+            check(
+                first_version.major,
+                first_version.minor,
+                format!("the specified edition `{}`", self.edition),
+            )?;
+        }
+
+        if let Some(profiles) = &self.profiles {
```

Comment:
Style nit: for profiles and features, I'd probably try to figure out if there's way to not have so much rightward drift. Maybe `self.profiles().as_deref().unwrap_or(&[])` and `self.original.features().map(|(_, features)| &features).unwrap_or(&[])`?